segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/SegmentSortedKeyIndex.java

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.tables;
+
+import com.google.common.annotations.Beta;
+import io.pravega.common.util.ArrayView;
+import io.pravega.common.util.AsyncIterator;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nullable;
+import lombok.Data;
+
+/**
+ * Defines an index that maintains a Table Segment's Keys in lexicographic bitwise order.
+ */
+@Beta
+interface SegmentSortedKeyIndex {
+    /**
+     * Include and persist updates that have been included in a Table Segment Index.
+     *
+     * @param bucketUpdates A Collection of {@link BucketUpdate} instances that reflect what keys have been added and/or
+     *                      removed.
+     * @param timeout       Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the operation has completed.
+     */
+    CompletableFuture<Void> persistUpdate(Collection<BucketUpdate> bucketUpdates, Duration timeout);
+
+    /**
+     * Includes the given {@link TableKeyBatch} which contains Keys that have recently been updated or removed, but not
+     * yet indexed. These will be stored in a memory data structure until {@link #updateSegmentIndexOffset} will be invoked
+     * with an offset that exceeds their offset.
+     *
+     * @param batch              The {@link TableKeyBatch} to include.
+     * @param batchSegmentOffset The offset of the {@link TableKeyBatch}.
+     */
+    void includeTailUpdate(TableKeyBatch batch, long batchSegmentOffset);
```

Comment:
The next three calls are to maintain a tail cache, if I understand it right. By reading these, it is not clear to me the expected flow of these calls for a user of this interface, and it is also not not clear what the consistency guarantee is.

Reply:
This mirrors `ContainerKeyIndex` and `ContainerKeyCache`. 
- `includeTailUpdate` is called every time we insert, update or remove a key (from the API)
- `includeTailCache` is called (optionally) after a container failover, if someone wants to access the Segment **prior to** it being fully indexed by the `WriterTableProcessor` (via `persistUpdate`. 
    - Why is this important? Because the in-memory part is volatile; it is gone after a crash. We accumulate data into it via `includeTailUpdate`, but that information is gone after a crash. There is special code in the `ContainerKeyIndex` (pre-existing; not part of this PR) that auto-recovers a Table Segment if it is accessed immediately after a recovery but before it is indexed. (If you remember, we had a few problems last year when the Controller was complaining that Table Segment data was unavailable shortly after a recovery - this pre-recovery was introduced to handle that kind of problems).
- `updateSegmentIndexOffset` is called by the `WriterTableProcessor` (indirectly, via `ContainerKeyIndex` to tell the SKI that everything prior to a certain offset has been indexed and it can remove it from the "tail cache" (i.e., if anyone needs that info, they can get it from the actual index, and no need to keep a heap data structure filled with redundant info).


Does this explanation make sense? These are all pre-existing concepts, and have just been applied to the SKI. The consumers of this interface are the `ContainerKeyIndex`, `WriterTableProcessor` which work in tandem to offer a unified view of the TableSegment to the upstream code.

Reply:
What actually got me to ask about it is whether this distinction between cached vs. indexed  belongs in the interface or whether it is an implementation detail leaking to the interface. But, if this is symmetric to an existing concept, then there is no point in arguing about it for this PR. At best, we should think about it separately.

Reply:
While this is an interface, it's an internal one, and the whole point of it was to be able to mock it during unit tests. This is buried so deep within the stack that it will be hard for anyone to surface this if they want to. 

Yes, it exposes details about the implementation, but this **is** the implementation.